bestiary/core/api.py

```diff
@@ -296,3 +314,74 @@ def delete_project(ctx, project_id):
     project.id = project_id
 
     return project
+
+
+@django.db.transaction.atomic
+def move_project(ctx, from_project_id, to_project_id):
+    """Move a project to a parent project and/or to an ecosystem.
+
+    Link a project with another one with a child-parent relation. If a project
+    is not linked to any project, it is a root project. Once the project is linked
+    with another one, it won't be a root project anymore.
+    If the source project was already linked to a parent project, it will be
+    overwritten in case a value is provided.
+
+    :param ctx: context from where this method is called
+    :param from_project_id: Project to be linked to another project or to an ecosystem
+    :param to_project_id: Destination project which will be linked as a parent project
+
+    :returns: the updated project object
+
+    :raises InvalidValueError: raised when from_project_id is None or any of the
+        other IDs are invalid
+    :raises NotFoundError: raised when the any of the projects or the ecosystem
+        do not exist in the registry
+    :raises InvalidParentError: raised when the parent project to be set
+        is invalid
+    """
+    def _get_root(project):
+        """Return the root project given one of its subprojects"""
+
+        parent = project.parent_project if project else None
+        project = _get_root(parent) if parent else project
+
+        return project
+    def _check_descendants(trxl, project, parent):
+        """Check if the parent project to be set is among its descendants"""
+
+        for subproject in project.subprojects.all():
+            if subproject == parent:
+                raise InvalidParentError(msg='A child project cannot be set as parent')
+            _check_descendants(trxl, subproject, parent)
```

Comment:
I think it is better to have a function that would say if a project is the child of another project. Also, try to avoid recursion. I don't think it make sense to use it in this case where you can use BFS or DFS algorithms.